src/sdam/server.ts

```
@@ -428,6 +419,73 @@ export class Server extends TypedEventEmitter<ServerEvents> {
     }
   }
 
+  /**
+   * Ensure that error is properly decorated and internal state is updated before throwing
+   * @internal
+   */
+  private decorateAndHandleError(
+    connection: Connection,
+    cmd: Document,
+    options: CommandOptions | GetMoreOptions | undefined,
+    error: AnyError
+  ): AnyError {
+    const session = options?.session;
+    if (error.name === 'AbortError' && error.cause instanceof MongoError) {
+      error = error.cause;
+    }
+
+    if (!(error instanceof MongoError)) {
+      // Node.js or some other error we have not special handling for
+      throw error;
+    }
+
+    if (connectionIsStale(this.pool, connection)) {
+      throw error;
+    }
+
+    if (error instanceof MongoNetworkError) {
+      if (session && !session.hasEnded && session.serverSession) {
+        session.serverSession.isDirty = true;
+      }
+
+      // inActiveTransaction check handles commit and abort.
+      if (
+        inActiveTransaction(session, cmd) &&
+        !error.hasErrorLabel(MongoErrorLabel.TransientTransactionError)
+      ) {
+        error.addErrorLabel(MongoErrorLabel.TransientTransactionError);
+      }
+
+      if (
+        (isRetryableWritesEnabled(this.topology) || isTransactionCommand(cmd)) &&
+        supportsRetryableWrites(this) &&
+        !inActiveTransaction(session, cmd)
+      ) {
+        error.addErrorLabel(MongoErrorLabel.RetryableWriteError);
+      }
+    } else {
+      if (
+        (isRetryableWritesEnabled(this.topology) || isTransactionCommand(cmd)) &&
+        needsRetryableWriteLabel(error, maxWireVersion(this)) &&
+        !inActiveTransaction(session, cmd)
+      ) {
+        error.addErrorLabel(MongoErrorLabel.RetryableWriteError);
+      }
+    }
+
+    if (
+      session &&
+      session.isPinned &&
+      error.hasErrorLabel(MongoErrorLabel.TransientTransactionError)
+    ) {
+      session.unpin({ force: true });
+    }
+
+    this.handleError(error, connection);
+
+    return error;
```

can we always "throw" instead of sometimes returning an error?